File: frontend/src/types/cost_code.ts
export interface CostCode {
  cost_code_id: number;
  division_number: string;
  division_name: string;
  category_number: string;
  category_name: string;
  subcategory_number: string;
  subcategory_name: string;
  item_code: string;
  item_name: string;
  income_account_number?: string;
  income_account_name?: string;
  expense_account_number?: string;
  expense_account_name?: string;
  default_unit_type?: string;
  default_unit_price?: number;
}

export interface CostCodeCreate {
  division_number: string;
  division_name: string;
  category_number: string;
  category_name: string;
  subcategory_number: string;
  subcategory_name: string;
  item_code: string;
  item_name: string;
  income_account_number?: string;
  income_account_name?: string;
  expense_account_number?: string;
  expense_account_name?: string;
  default_unit_type?: string;
  default_unit_price?: number;
}

export interface CostCodeTreeNode {
  division_number: string;
  division_name: string;
  categories: CategoryNode[];
}

export interface CategoryNode {
  category_number: string;
  category_name: string;
  subcategories: SubcategoryNode[];
}

export interface SubcategoryNode {
  subcategory_number: string;
  subcategory_name: string;
  items: CostCode[];
}

export interface CostCodeTree {
  divisions: CostCodeTreeNode[];
}

export type UnitType = 'gsf' | 'far' | 'gar_sf' | 'slab' | 'units' | 'months' | 'weeks' | 'days' | 'ls' | 'ea';

export const UNIT_TYPE_LABELS: Record<UnitType, string> = {
  gsf: 'Gross SF',
  far: 'FAR SF',
  gar_sf: 'Garage SF',
  slab: 'Slab SF',
  units: 'Units',
  months: 'Months',
  weeks: 'Weeks',
  days: 'Days',
  ls: 'Lump Sum',
  ea: 'Each',
};

File: backend/app/schemas/cost_code.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List


class CostCodeCreate(BaseModel):
    division_number: str
    division_name: str
    category_number: str
    category_name: str
    subcategory_number: str
    subcategory_name: str
    item_code: str
    item_name: str
    income_account_number: Optional[str] = None
    income_account_name: Optional[str] = None
    expense_account_number: Optional[str] = None
    expense_account_name: Optional[str] = None
    default_unit_type: Optional[str] = None
    default_unit_price: Optional[float] = None


class CostCodeUpdate(BaseModel):
    division_number: Optional[str] = None
    division_name: Optional[str] = None
    category_number: Optional[str] = None
    category_name: Optional[str] = None
    subcategory_number: Optional[str] = None
    subcategory_name: Optional[str] = None
    item_code: Optional[str] = None
    item_name: Optional[str] = None
    income_account_number: Optional[str] = None
    income_account_name: Optional[str] = None
    expense_account_number: Optional[str] = None
    expense_account_name: Optional[str] = None
    default_unit_type: Optional[str] = None
    default_unit_price: Optional[float] = None


class CostCodeResponse(BaseModel):
    cost_code_id: int
    division_number: str
    division_name: str
    category_number: str
    category_name: str
    subcategory_number: str
    subcategory_name: str
    item_code: str
    item_name: str
    income_account_number: Optional[str] = None
    income_account_name: Optional[str] = None
    expense_account_number: Optional[str] = None
    expense_account_name: Optional[str] = None
    default_unit_type: Optional[str] = None
    default_unit_price: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


class CostCodeTree(BaseModel):
    cost_code_id: int
    division_number: str
    division_name: str
    category_number: str
    category_name: str
    subcategory_number: str
    subcategory_name: str
    item_code: str
    item_name: str
    income_account_number: Optional[str] = None
    income_account_name: Optional[str] = None
    expense_account_number: Optional[str] = None
    expense_account_name: Optional[str] = None
    default_unit_type: Optional[str] = None
    default_unit_price: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)
